### PR TITLE
Bump cardano-node to (almost) 1.28.0

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

ADP-952

# Overview

- [x] Bump to cardano-node master, in preparation for 1.28.0 release (input-output-hk/cardano-haskell#49)
- [x] Update readme 

# Comments

Basic Alonzo Era support is a non-goal for this PR. I implemented `fromAlonzoTx` and `fromAlonzoBlock` as `error "unimplemented"`s for now. Will fill them in in a follow-up PR, perhaps with integration cluster support at the same time.

- [Hydra jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet-pr-2755#tabs-evaluations)
- [Cabal build](https://buildkite.com/input-output-hk/cardano-wallet-nightly/builds?branch=anviking%2FADP-952%2Fcardano-node-1.28.0)

### CODE REVIEW

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 1]
```diff
@@ -1191,36 +1207,42 @@ mkUnsignedTx era ttl cs md wdrls certs fees =
         ShelleyBasedEraShelley -> toShelleyTxOut
         ShelleyBasedEraAllegra -> toAllegraTxOut
         ShelleyBasedEraMary -> toMaryTxOut
+        ShelleyBasedEraAlonzo -> error "todo: toAlonzoTxOut"
```
Comment (Anviking) [inline lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]:
```suggestion
        ShelleyBasedEraAlonzo -> error "todo(ADP-952): toAlonzoTxOut"
```

**Review by rvl — APPROVED** [review]

**Comment by Anviking** [comment]
I now get
```
plutus-core/src/PlutusCore/Default/Builtins.hs:123:9: error:
--
  | • Could not deduce (GEq DefaultUni)
  | arising from a use of ‘makeBuiltinMeaning’
  | from the context: uni ~ DefaultUni
  | bound by the instance declaration
  | at plutus-core/src/PlutusCore/Default/Builtins.hs:120:10-60
  | or from: PlutusCore.Constant.Typed.HasConstantIn uni term
  | bound by the type signature for:
  | toBuiltinMeaning :: forall term.
  | PlutusCore.Constant.Typed.HasConstantIn uni term =>
  | DefaultFun
  | -> BuiltinMeaning term (CostingPart uni DefaultFun)
  | at plutus-core/src/PlutusCore/Default/Builtins.hs:122:5-20
  | • In the expression:
  | makeBuiltinMeaning
  | ((+) @Integer) (runCostingFunTwoArguments . paramAddInteger)
  | In an equation for ‘toBuiltinMeaning’:
  | toBuiltinMeaning AddInteger
  | = makeBuiltinMeaning
  | ((+) @Integer) (runCostingFunTwoArguments . paramAddInteger)
  | In the instance declaration for ‘ToBuiltinMeaning uni DefaultFun’
  | \|
  | 123 \|         makeBuiltinMeaning
```
for both stack and cabal, even though it seemed fixed locally (for stack build at least) 🤷‍♂️

File: lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs (hunk 1 above)
Comment (rvl) [reply lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs]: If you use something like this:
```haskell
error "toAlonzoTxOut unimplemented" -- TODO: [ADP-952] toAlonzoTxOut
```
then it's easy to find all your TODOs with lentil:
```shell
$ lentil -t ADP-952 lib
263 source files [****************] 100%
lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs
   521  fromAlonzoBlock [ADP-952]
   923  fromAlonzoTx [ADP-952]
  1148  datumHash [ADP-952]
  1167  datumHash [ADP-952]
  1186  datumHash [ADP-952]
 
lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs
  1173  We presumably need to provide the protocol params if our tx uses
        scripts? [ADP-952]
  1220  toAlonzoTxOut [ADP-952]
```

**Comment by rvl** [comment]
Thanks @Anviking - I have done a squash and rebase.

The stack build error on Buildkite was because the auto-generated nix files weren't up-to-date, and this was because it failed to push back to github for some weird reason. But anyway, I regenerated them.

Let's see what happens with the integration tests...

bors r+

**Comment by rvl** [comment]
Again

bors r+

**Comment by Anviking** [comment]
bors try

**Comment by Anviking** [comment]
> The stack build error on Buildkite was because the auto-generated nix files weren't up-to-date, and this was because it failed to push back to github for some weird reason. But anyway, I regenerated them.

Many thanks!

Not sure if I should squash everything, but merging.
bors r+

**Comment by Anviking** [comment]
_Fixed readme node revision and squashed_

bors r+

**Comment by Anviking** [comment]
bors cancel

**Comment by Anviking** [comment]
bors r-

Saw the PastHorizon failure again...

**Comment by Anviking** [comment]
bors try

**Comment by Anviking** [comment]
What is strange about https://hydra.iohk.io/build/6897950/nixlog/1 is that the PastHorizon failure come from reading a wallet in a part of a tests which doesn't use wallets. What actually caused the failure was CLI_NETWORK, which provided little details as of why it failed.

In https://buildkite.com/input-output-hk/cardano-wallet/builds/15481#6b969a8a-2a4a-4947-a21f-c1b41bae0882, there were PastHorizonErrors, but the retrying `it` ultimately made the test pass.

I can get PastHorizon errors locally using with `stack test cardano-wallet:integration --ta '-j 25'.`

I think there might be nothing in particular going on except that the tip-sync client had to compete with wallet chain-sync clients, and got out of date.

### Things that might help down the line:
1. Use one chain-sync client for multiple wallets to reduce load
2. Chain-sync refactoring to handle rollbacks without re-negotiating (might reduce load)
3. Add logic to prioritize tip-sync client above wallets
5. Somehow get the TimeInterpreter from chain-sync, e.g. https://github.com/input-output-hk/ouroboros-network/pull/2901

### Things we could do now:
1. Do nothing and merge for now
2. Reduce parallelism in CI and/or increase slot or epoch length
3. Attempt to recover from PastHorizon errors by waiting for the interpreter to be such up to date that all `Tip block` we have seen (from all chain-sync threads) are within the horizon.
    - Obviously more tricky than the above, but might be doable 🤔

**Pull request merged**